Userland/Libraries/LibELF/DynamicLinker.cpp

```diff
@@ -237,42 +247,50 @@ static Result<void, DlErrorMessage> map_dependencies(ByteString const& path)
     return {};
 }
 
-struct ThreadSpecificData {
-    ThreadSpecificData* self;
-};
-
 static ErrorOr<FlatPtr> __create_new_tls_region()
 {
-    auto static_tls_region_alignment = max(s_tls_data.alignment, alignof(ThreadSpecificData));
-    auto static_tls_region_size = align_up_to(s_tls_data.tls_template_size, static_tls_region_alignment) + sizeof(ThreadSpecificData);
-    void* thread_specific_ptr = serenity_mmap(nullptr, static_tls_region_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0, static_tls_region_alignment, "Static TLS Data");
-    if (thread_specific_ptr == MAP_FAILED)
+    void* static_tls_region = serenity_mmap(nullptr, s_tls_data.static_tls_region_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0, s_tls_data.static_tls_region_alignment, "Static TLS Data");
+    if (static_tls_region == MAP_FAILED)
         return Error::from_syscall("mmap"sv, -errno);
 
-    auto* thread_specific_data = bit_cast<ThreadSpecificData*>(bit_cast<FlatPtr>(thread_specific_ptr) + (align_up_to(s_tls_data.tls_template_size, static_tls_region_alignment)));
-    thread_specific_data->self = thread_specific_data;
+    auto thread_pointer = calculate_tp_value_from_static_tls_region_address(bit_cast<FlatPtr>(static_tls_region), s_tls_data.tls_template_size, s_tls_data.static_tls_region_alignment);
+    VERIFY(thread_pointer % s_tls_data.static_tls_region_alignment == 0);
 
-    auto* thread_local_storage = bit_cast<u8*>(bit_cast<FlatPtr>(thread_specific_data) - align_up_to(s_tls_data.tls_template_size, s_tls_data.alignment));
+    auto* tcb = get_tcb_pointer_from_thread_pointer(thread_pointer);
+
+    // FIXME: Add support for dynamically-allocated TLS blocks.
+    tcb->dynamic_thread_vector = nullptr;
+
+#if ARCH(X86_64)
+    tcb->thread_pointer = bit_cast<void*>(thread_pointer);
+#endif
```

Comment:
should this check `if constexpr (TLS_VARIANT == 2)` instead?

Reply:
No, only x86-64 needs this quirk of having a copy of the thread pointer at offset 0, as you otherwise can't easily get the value of the thread pointer.
I also described that quirk in x86-64's tls.h.